### PR TITLE
chore: observability and security for HTTP gateway

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

Currently, to handle HTTP requests in Evonode we use outdated version of Envoy This version contains [many security vulnerabilities](https://www.cvedetails.com/vulnerability-list/vendor_id-19794/product_id-53798/opdos-1/Envoyproxy-Envoy.html) and is not configured properly for being an edge proxy for all incoming HTTP traffic. To control and limit the load it should provide mechanisms to limit incoming requests. We don't have any visibility on HTTP traffic in general and specific requests to Drive and DAPI.

## What was done?

- Updated Envoy and configuration to latest version 1.30.1
- Renamed DAPI Envoy service to Platform Gateway since it handles requests not just for DAPI but for Drive as well
- Added configuration `platform.gateway.maxConnections` to limit max active connections to Gateway. If we close to the limit, we disable keep-alive, and stop accepting requests when we reach it.
- Added configuration `platform.gateway.maxHeapSizeInBytes` to release free memory and disable keep-alive when we close to limit and reject incoming connections when we reach it.
- Added configuration `platform.gateway.upstreams.*.maxRequest` to limit active requests to specific upstream (DAPI, Drive, etc) 
- Added configuration `platform.gateway.metrics` to expose Envoy Prometheus metrics.
- Added configuration `platform.gateway.admin` to enable Envoy admin interface (must be enabled for metrics).
- Added configuration `platform.gateway.listeners.*.http2.maxConcurrentStreams` to limit max concurrent streams per incoming HTTP2 connection.
- Added configuration `platform.gateway.log.level` to define verbosity of Envoy application logs.
- Added configuration `platform.gateway.log.accessLogs` to define any number of access logs. Supported destinations are `stdout`, `stderr` and `file`. Supported formats are `text` or `json`.  Default output templates are defined for each format and can be ovewritten for any specific access logs.
- Global HTTP request rate limiter replaced with IP-based so now it limits per IP but not for all incoming requests. Whitelisting and blacklisting are now supported as well.
- Added configuration `platfrom.gateway.rateLimiter.metrics` to expose Rate limiter Prometheus metrics.
- Added configuration `platform.drive.abci.metrics` to expose Drive ABCI Prometheus metrics.
- the `dapi_tx_filter_stream` docker compose service renamed to `dapi_core_streams` since it expose multiple streaming endpoints.
- Configured Envoy to sanitize incoming request path
- Configured Envoy to limit the connection and request max duration and idle timeouts
- Configured Envoy to limit http2 connection and stream window sizes 
- Rejected external requests to `getProofs` since this is internal endpoint between DAPI and Drive and can be misused to make a huge load on the system.
- Reduced new connection read and write buffers to reduce memory consumption
- Forced more efficient HTTP2 connections between Envoy and upstreams
- Configured Envoy to expose Prometheus metrics with default path instead of `/stats/prometheus`
- Added metrics for Drive ABCI queries 

## How Has This Been Tested?

Multiple testing scenarios to verify each configuration on the ouzo devnet with metric visualisations in Grafana

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

File: packages/rs-drive-abci/.env.testnet [hunk 1]
```diff
@@ -1,6 +1,6 @@
 # ABCI host and port to listen
 ABCI_CONSENSUS_BIND_ADDRESS="tcp://0.0.0.0:26658"
-PROMETHEUS_BIND_ADDRESS="http://0.0.0.0:29090"
+#PROMETHEUS_BIND_ADDRESS="http://0.0.0.0:29090"
```
Comment (shumkov) [inline packages/rs-drive-abci/.env.testnet:3]: Comment

File: packages/rs-drive-abci/src/logging/level.rs [hunk 2]
```diff
@@ -110,7 +110,7 @@ impl TryFrom<&LogLevel> for EnvFilter {
                     .expect("should be a valid log specification")
             }
             LogLevel::Debug => EnvFilter::try_new(
-                "info,tenderdash_abci=debug,drive_abci=debug,drive=debug,dpp=debug",
+                "info,tenderdash_abci=debug,drive_abci=debug,drive=debug,dpp=debug",  // TODO: Add tonic
```
Comment (shumkov) [inline packages/rs-drive-abci/src/logging/level.rs:113]: Todo

File: packages/dashmate/templates/platform/gateway/envoy.yaml.dot [hunk 3]
```diff
@@ -0,0 +1,486 @@
+!ignore filters: &filters
+  - name: envoy.http_connection_manager
```
Comment (shumkov) [inline packages/dashmate/templates/platform/gateway/envoy.yaml.dot:2]: Shall we rate limit incoming connections by IP? RIght now we limit only incoming requests by IP.

Comment (shumkov) [reply packages/dashmate/templates/platform/gateway/envoy.yaml.dot:2]: We agreed that this must be done on host level or with external services and we can't do anything with this at dashmate level.

File: packages/dashmate/configs/defaults/getBaseConfigFactory.js [hunk 4]
```diff
@@ -125,34 +125,85 @@ export default function getBaseConfigFactory(homeDir) {
...
       platform: {
-        dapi: {
-          envoy: {
-            docker: {
-              image: 'dashpay/envoy:1.22.11',
+        gateway: {
+          docker: {
+            image: 'dashpay/envoy:1.30.2-impr.1',
+          },
+          maxConnections: 10000,
```
Comment (lklimek) [inline packages/dashmate/configs/defaults/getBaseConfigFactory.js]: how does it relate to max no of open files?

and 10 000 looks way too much.

File: packages/dashmate/configs/defaults/getBaseConfigFactory.js [hunk 5]
```diff
@@ -125,34 +125,85 @@ export default function getBaseConfigFactory(homeDir) {
...
+          maxHeapSizeInBytes: 125000000, // 1 Gb
+          upstreams: {
+            driveGrpc: {
+              maxRequests: 5000,
```
Comment (lklimek) [inline packages/dashmate/configs/defaults/getBaseConfigFactory.js]: 100

File: packages/dashmate/src/config/Config.js [hunk 6]
```diff
@@ -182,4 +182,4 @@ export default class Config {
 }
 
 Config.ajv = new Ajv({ coerceTypes: true });
-addFormats(Config.ajv, { mode: 'fast', formats: ['ipv4'] });
+addFormats(Config.ajv, { mode: 'fast', formats: ['ipv4', 'ipv6'] });
```
Comment (lklimek) [inline packages/dashmate/src/config/Config.js]: remove ipv6

File: packages/rs-drive-abci/src/metrics.rs [hunk 7]
```diff
@@ -170,7 +191,24 @@ impl Prometheus {
...
+                metrics::Unit::Seconds,
                 "Duration of ABCI request execution inside Drive per endpoint, in seconds"
+            );
+
+            describe_counter!(
+                COUNTER_LAST_HEIGHT,
+                "Last finalized height of platform chain (eg. Tenderdash)"
+            );
+
+            describe_counter!(
```
Comment (lklimek) [inline packages/rs-drive-abci/src/metrics.rs]: duplicate?

File: packages/rs-drive-abci/src/query/service.rs [hunk 8]
```diff
@@ -151,7 +163,65 @@ impl QueryService {
...
+                    "query '{}' executed with code {:?} in {} secs",
+                    endpoint_name,
+                    code,
+                    elapsed_time
+                );
+            }
+            // System errors
+            Code::Cancelled
+            | Code::Unknown
+            | Code::DeadlineExceeded
```
Comment (lklimek) [inline packages/rs-drive-abci/src/query/service.rs:208]: looks like it's user error, same as cancelled

**Review by lklimek — APPROVED** [review]

**Review by lklimek — APPROVED** [review]

**Review by QuantumExplorer — APPROVED** [review]
Approving without review.

**Pull request merged**